### DISCUSSION TITLE
MtA (Multiplicative to Additive) share conversion  protocol for Ed25519

### DISCUSSION THREAD

**Discussion by HRezaei** [ZenGo-X/multi-party-ecdsa#157] [Q&A]
https://github.com/ZenGo-X/multi-party-ecdsa/blob/fc4a0ed2e1c9ba75d249e7ef0172ebc310f2eb43/src/utilities/mta/test.rs#L7

This test function works fine with the secp256_k1 curve, but if you change the underlying curve scalars and points to those of Ed25519, it fails always. To see the issue you need to refactor the test function and also `MessageA`, `MessageB`, etc. You can take a look at [this repo](https://github.com/HRezaei/MtA-Test) where I prepared the mta_test in isolation and performed tests over all the curves (using curv-kzen v0.9). The result shows issues with Ed25519 and Ristretto, others curves work fine.

**Comment by elichai** [comment]
Yes, the range proofs assume it is secp256k1, See: https://github.com/ZenGo-X/multi-party-ecdsa/blob/master/src/utilities/mta/range_proofs.rs#L153

**Reply by HRezaei** [reply]
Thanks for the reply.

So what happens when we want to implement gg2018 signing protocol for EdDSA? We need MtA check, don't we?

**Reply by survived** [reply]
> gg2018 signing protocol for EdDSA

Do you mean ECDSA instantiated with ed25519 curve? EDDSA is a another signing scheme

**Reply by survived** [reply]
> Yes, the range proofs assume it is secp256k1

@elichai, is it by construction, or it can be updated to work with any curve? I see only one place where secp256k1 is used: [here](https://github.com/ZenGo-X/multi-party-ecdsa/blob/fc4a0ed2e1c9ba75d249e7ef0172ebc310f2eb43/src/utilities/mta/range_proofs.rs#L161). @HRezaei has updated this piece of code ([here](https://github.com/HRezaei/MtA-Test/blob/e88e2698c00f2f0a048b1e830cfeb331ec9b637d/src/utilities/mta/range_proofs.rs#L171)), but MtA doesn't work anyway

**Reply by DmytroTym** [reply]
@survived I think it should be possible to update it to work with any curve. I'll look into it

**Reply by DmytroTym** [reply]
PS. @HRezaei do groups of curves that you're trying to use have order approximately `2^256`? It's important that the output of hash function, which is `Sha256` here, is (at least approximately) the same as the order of your EC group.

**Reply by DmytroTym** [reply]
For example [here](https://github.com/HRezaei/MtA-Test/blob/e88e2698c00f2f0a048b1e830cfeb331ec9b637d/src/utilities/mta/range_proofs.rs#L525), note that we treat `e` as a `Scalar::<E>`, where `e` is an output of `Sha256`

**Reply by HRezaei** [reply]
> do groups of curves that you're trying to use have order approximately 2^256? 

Yes, AFAIK. We particularly are interested in Ed25519 which I think has the order of 2^256. 

But in general, the [tests here](https://github.com/HRezaei/MtA-Test) are run over all curves you have defined in curv-kzen v0.9 and all are OK but Ed25519 and Ristretto.

**Reply by elichai** [reply]
But why would you want to do Ed25519 with this?
Ed25519 is a linear signature scheme and you could use FROST, MuSig2 and other simpler schemes than gg

**Reply by DmytroTym** [reply]
> But in general, the tests here are run over all curves you have defined in curv-kzen v0.9 and all are OK but Ed25519 and Ristretto.

Yeah, I ran the tests locally. For `ed25519` the problem is indeed somewhere around converting the output of `Sha256` into a `Scalar::<E>`. I don't know too much about this curve, or how it's implemented, but in order for MtA to work we need the following property of converting `BigInt`s to `Scalar::<E>`: if `a` and `b` are `BigInt`, then `Scalar::<E>::from(a)+Scalar::<E>::from(b)=Scalar::<E>::from(a+b)`. This seems to break for `ed25519`.
For `ristretto`, the method `to_x_coord()`, called on a point returns `None` instead of a `BigInt` value [here](https://github.com/HRezaei/MtA-Test/blob/master/src/utilities/mta/range_proofs.rs#L453). Tbh I have no idea what `ristretto` is, but I guess the fix is to change `to_x_coord()` method to anything that takes a point and returns a `BigInt` value crafted from that point.

**Reply by HRezaei** [reply]
> Do you mean ECDSA instantiated with ed25519 curve? EDDSA is a another signing scheme

I used to think they are the same! Now with the light you shed, I would say I need EdDSA. You said it is different from ECDSA instantiated with ed25519. Does this mean that MtA is not necessary for EdDSA?

> But why would you want to do Ed25519 with this?

 I want to implement {t,n} threshold EdDSA, thus I tried to implement something similar to [gg2018_sign_client](https://github.com/ZenGo-X/multi-party-ecdsa/blob/master/examples/gg18_sign_client.rs) using Ed25519 instead of Secp256k1. And it always fails in the [MtA check step](https://github.com/ZenGo-X/multi-party-ecdsa/blob/fc4a0ed2e1c9ba75d249e7ef0172ebc310f2eb43/examples/gg18_sign_client.rs#L226).

**Reply by HRezaei** [reply]
More broadly, we need {t,n} threshold EdDSA for signing Solana or Stellar transactions.

**Reply by elichai** [reply]
@HRezaei Maybe look here instead? https://github.com/zenGo-X/multi-party-eddsa

**Reply by HRezaei** [reply]
@elichai: Yeah, I did. Indeed, I ported gg18 examples into the repo you mentioned.

**Reply by DmytroTym** [reply]
> More broadly, we need {t,n} threshold EdDSA for signing Solana or Stellar transactions.

Yep, you definitely rather need https://github.com/zenGo-X/multi-party-eddsa. Also, afaik, its simpler and faster than GG18 or GG20.

**Reply by elichai** [reply]
Yes, It's at least a couple of orders of magnitude simpler :)
also, There's also a great implementation of FROST here if you want Ristretto: https://docs.rs/frost-dalek/latest/frost_dalek/
